src/protocols/bdx/BDXMessageUtils.h

```diff
@@ -0,0 +1,231 @@
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines structures and utility methods for working with BDX
+ *      messages.
+ */
+
+#pragma once
+
+#include <support/CodeUtils.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace BDX {
+
+enum ControlMode : uint8_t
+{
+    kNotSpecified  = 0x00,
+    kSenderDrive   = 0x10,
+    kReceiverDrive = 0x20,
+    kAsync         = 0x40,
+};
+
+/*
+ * A structure for representing a SendInit or ReceiveInit message (both contain
+ * identical parameters).
+ */
+struct TransferInit
+{
+    /**
+     * @brief
+     *  Pack (write) the message into a PacketBuffer.
+     *
+     * @param[out] aBuffer a PacketBuffer to use to write the message
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to write to the PacketBuffer
+     */
+    CHIP_ERROR Pack(System::PacketBuffer & aBuffer);
+
+    /**
+     * @brief
+     *  Parse data from an PacketBuffer into a struct instance
+     *
+     * @param[in] aBuffer Pointer to a PacketBuffer containing the data.
+     * @param[out] aParsedMessage Reference to a struct instance where results will be stored.
+     *             Note that this struct will store pointers into the passed PacketBuffer, so it
+     *             is essential that the PacketBuffer is not modified or freed until after the
+     *             TransferInit struct is no longer needed.
+     *
+     * @return CHIP_ERROR Any error that occurs when trying to read the message
+     */
+    static CHIP_ERROR Parse(System::PacketBuffer & aBuffer, TransferInit & aParsedMessage);
+
+    /**
+     * @brief
+     *  Get the size of the message once it is written to a buffer.
+     */
+    size_t PackedSize();
+
+    /**
+     * @brief
+     *  Equality check method.
+     */
+    bool operator==(const TransferInit &) const;
+
+    // Proposed Transfer Control (required)
+    bool mSupportsAsync;
+    bool mSupportsReceiverDrive;
+    bool mSupportsSenderDrive;
+    uint8_t mSupportedVersions;
+
+    // Range Control (required)
+    bool mWideRange;       ///< Set true to indicate mStartOffset and mDefLen are 64-bit, else 32-bit.
+    uint64_t mStartOffset; ///< Proposed start offset of data. 0 for no offset. Will be truncated if mWideRange is false.
+    bool mDefLen;          ///< True if transfer has a definite length.
+
+    uint64_t mMaxLength;    ///< Proposed max length of data in transfer, 0 for indefinite (required if mDefLen is set).
+                            ///< Will be truncated if mWideRange is false.
+    uint16_t mMaxBlockSize; ///< Proposed max block size to use in transfer (required)
+
+    // File designator (required)
+    uint8_t * mFileDesignator;
+    uint16_t mFileDesLength; ///< Length of file designator string (not including null-terminator)
+
+    // Additional metadata (optional, TLV format)
+    uint8_t * mMetadata;
+    uint16_t mMetadataLength;
+};
+
+struct SendInit : public TransferInit
```

Comment:
do we expect these to change or could we start as a:

`using SendInit = TransferInit;` 

? Is there value in having completely separate types?

Reply:
Oh, I've never seen that syntax before. Would you suggest putting that in the header? 

I don't think we expect these messages to have different fields right now. @tcarmelveilleux - do you expect us to add any unique fields to SendInit or ReceiveInit in the future?

Reply:
Yes, the `using` clauses should be in the header as they alias types.

Because it's already split in the SendInit/TransferInit name, if we do add fields we could introduce separate structs/classes in the future.

Reply:
Sounds good, I'll replace with `using`